### PR TITLE
[NEW] nginx-ingress

### PR DESCRIPTION
**Pull request by mgoodness**
This is a stand-alone chart for the nginx ingress controller that, along with the kube-lego chart (#397), replaces nginx-lego. As mentioned and (briefly) discussed in #385.

@jackzampolin

### CODE REVIEW

**Comment by jackzampolin** [comment]
@mgoodness This looks great! I'd be happy to stay on as the maintainer! 😉  It's much better as two charts. I'll pull this down and give it a test.

**Comment by viglesiasce** [comment]
Tested this over the weekend without any issues.

Will wait for @jackzampolin to give the LGTM

**Comment by bacongobbler** [comment]
ping @jackzampolin, does this have your LGTM?

**Comment by jackzampolin** [comment]
@bacongobbler Got to this today. LGTM. Sorry for the delay there.

**Comment by prydonius** [comment]
This is the first time we're changing the name/deprecating a chart, since nginx-lego will stay in the repo, is there something we can do to let people know it's deprecated? e.g. we could push a final version that explains it's moved to nginx-ingress in the NOTES.txt?

**Comment by jackzampolin** [comment]
@prydonius I don't know how many users it had. In this particular case there was an issue (my b) with the other chart that prevented it from deploying successfully. 

It would be nice to do in the future but I think it might be overkill for this particular instance.

**Comment by jason-riddle** [comment]
I ran into a small hiccup in regards to the instructions.

```
export NODE_PORT=$(kubectl --namespace default get services -o jsonpath="{.spec.ports[0].nodePort}" silly-aardvark-nginx-ingress-controller)

export NODE_IP=$(kubectl --namespace default get nodes -o jsonpath="{.items[0].status.addresses[0].address}")

echo "Visit http://$NODE_IP:$NODE_PORT to access your application."
```

~~~I examined the NODE_IP which appears to be using the internal cluster ip. I believe this is due to the fact that I applied this chart with~~~

```
  services:
    - name: controller
      # annotations:
      type: NodePort
```

~~So after I created my ingress, I queried it and it appear to has the correct external facing ip.~~

Update below.

**Comment by jason-riddle** [comment]
Update: I mispoke. The problem for me was 

```
kubectl --namespace default get nodes -o jsonpath="{.items[0].status.addresses[0].address}"
```

should be

```
kubectl --namespace default get nodes -o jsonpath="{.items[0].status.addresses[1].address}"
```

so

```bash
status.addresses[1]
# instead of status.addresses[0]
```

**Comment by prydonius** [comment]
@jason-riddle it appears that a lot of charts use this in their NOTES, can you create a separate issue for it?

**Comment by jason-riddle** [comment]
sure

**Comment by munnerz** [comment]
Hey @mgoodness - I've just opened https://github.com/kubernetes/charts/pull/540, and then gone on to spot this PR :)

You seem to have put more into documenting so think it's best we stick with your PR, but may I request you add support for TCP & UDP configmaps into yours too? You can see an example of how I've done it in the above linked PR. I'm happy to do it myself, but will need to wait until after this has merged if so!

**Comment by viglesiasce** [comment]
@k8s-bot e2e test this

File: stable/nginx-ingress/templates/controller-service.yaml
Comment (viglesiasce) [inline stable/nginx-ingress/templates/controller-service.yaml]: Tests are failing because of this. Do we have to loop in this file? When would I have multiple? Can we start with this hardcoded to 1 service?

**Review by viglesiasce — DISMISSED** [review]

File: stable/nginx-ingress/templates/controller-service.yaml
Comment (mgoodness) [reply stable/nginx-ingress/templates/controller-service.yaml]: My specific use-case is to have the same nginx-ingress controllers exposed via two different ELBs (one internet-facing and one VPC-internal). But now that I think about it, that shouldn't be strictly necessary. I'll play with it over the weekend.

**Comment by mgoodness** [comment]
Addressed comments & requests from @viglesiasce, @jason-riddle, and @munnerz.

File: stable/nginx-ingress/templates/controller-daemonset.yaml
Comment (kfox1111) [inline stable/nginx-ingress/templates/controller-daemonset.yaml:23]: Needs an optional nodeSelector section from values.

Otherwise, it may land ingress controllers on unsuitable nodes.

File: stable/nginx-ingress/templates/controller-deployment.yaml
Comment (kfox1111) [inline stable/nginx-ingress/templates/controller-deployment.yaml:24]: Optional nodeSelector.

File: stable/nginx-lego/templates/default-deployment.yaml
Comment (kfox1111) [inline stable/nginx-lego/templates/default-deployment.yaml]: nodeSelector

Could this be added as a sidecar to help monitor easier?
https://github.com/hnlq715/nginx-vts-exporter

File: stable/nginx-ingress/templates/nginx-deployment.yaml
Comment (kfox1111) [inline stable/nginx-ingress/templates/nginx-deployment.yaml]: Could you add this as a sidecar to make monitoring easier?
https://github.com/hnlq715/nginx-vts-exporter

File: stable/nginx-lego/templates/nginx-deployment.yaml
Comment (kfox1111) [inline stable/nginx-lego/templates/nginx-deployment.yaml]: https://github.com/hnlq715/nginx-vts-exporter 

nodeSelector

File: stable/nginx-ingress/templates/default-deployment.yaml
Comment (kfox1111) [inline stable/nginx-ingress/templates/default-deployment.yaml]: Needs an optional nodeSelector to restrict where it lands in non uniform clusters.

File: stable/nginx-ingress/templates/nginx-service.yaml
Comment (kfox1111) [inline stable/nginx-ingress/templates/nginx-service.yaml]: This needs an optional externalIPs section. This would allow it to easily be able to be bound to vips for clusters that use those.

**Review by kfox1111 — CHANGES_REQUESTED** [review]
This looks really good. I went to test it on my cluster but ran into a few small issues preventing deployment in my environment. Specifically with not being able to pin it to some node types and the lack of externalIPs on the main service.

File: stable/nginx-ingress/templates/controller-service.yaml
Comment (unguiculus) [inline stable/nginx-ingress/templates/controller-service.yaml]: Should be `.Values.service.loadBalancerIP`

File: stable/nginx-ingress/templates/controller-service.yaml
Comment (unguiculus) [inline stable/nginx-ingress/templates/controller-service.yaml]: Should be `.Values.service.loadBalancerSourceRanges`

File: stable/nginx-ingress/templates/controller-service.yaml
Comment (unguiculus) [inline stable/nginx-ingress/templates/controller-service.yaml]: Should be `.Values.service.loadBalancerSourceRanges`

**Review by unguiculus — CHANGES_REQUESTED** [review]

File: stable/nginx-ingress/templates/default-backend-deployment.yaml
Comment (kfox1111) [inline stable/nginx-ingress/templates/default-backend-deployment.yaml:2]: Could you make this, and the default-backend-svc.yaml optional the same way the lego stuff was?

This would allow a user to load in their own default server, like a regular nginx to serve out a site customized/branded 404 page.

File: stable/nginx-lego/templates/default-deployment.yaml
Comment (mgoodness) [reply stable/nginx-lego/templates/default-deployment.yaml]: This file is being deleted.

File: stable/nginx-lego/templates/nginx-deployment.yaml
Comment (mgoodness) [reply stable/nginx-lego/templates/nginx-deployment.yaml]: This file is being deleted.

File: stable/nginx-ingress/templates/default-backend-deployment.yaml
Comment (mgoodness) [reply stable/nginx-ingress/templates/default-backend-deployment.yaml:2]: A user could do that using the current setup. Just need to set the `defaultBackend` values appropriately.

Comment (kfox1111) [reply stable/nginx-ingress/templates/default-backend-deployment.yaml:2]: Oh, I guess you could load a different image using the defaultBackend properties. So that might work for some use cases.

It would not give you the ability to load a differently setup pod though. So, like, if I wanted to load a deployment with an nginx container for my branded page, a fluentd sidecar to pull logs, and a prometheus exporter to monitor it.

So I still think the ability to disable its automatic creation is valuable?

File: stable/nginx-lego/templates/default-deployment.yaml
Comment (kfox1111) [reply stable/nginx-lego/templates/default-deployment.yaml]: Doh, your right. Sorry.

What do you think of the nginx-vtx-exporter sidecar though? Still applies to the new version too.

**Review by kfox1111 — CHANGES_REQUESTED** [review]

**Comment by mgoodness** [comment]
Not a huge fan of how I handled the default backend, so I'm open to suggestions. Will touch up the README tomorrow.

**Comment by kfox1111** [comment]
Thanks for fixing the lack of nodeSelector and adding the exporter sidecar.

I tested it out and its mostly working except I had to add the following to the controller-daemonset and deployment:

      annotations:
        "prometheus.io/scrape": {{ .Values.controller.enableStats | quote }}
        "prometheus.io/port": {{ .Values.statsExporter.service.containerPort | quote }}

A Prometheus instance I launched from the chart in this repo then automatically picked it up and started collecting metrics on all the instances without needing any configuration. Very cool. :)

Could you please add that bit to the pr?

**Comment by mgoodness** [comment]
Prometheus annotations (or any others) can now be listed in the `podAnnotations` values:
```
controller:
  podAnnotations:
    prometheus.io/scrape: "true"
    prometheus.io/port: "9913"
    ...
```
While these annotations are the ones recommended by Prometheus, they are not universal or mandatory. Users may use different ones for service discovery. I'm opting for flexibility.

**Comment by kfox1111** [comment]
Got a chance to play with this a bit more tonight.

I still need a way to set:
  externalIPs:
    - xxx.xxx.xxx.xxx
    - yyy.yyy.yyy.yyy
on the services.

Once I did that, I was able to bind the ingress to Prometheus and see its ui outside my baremetal cluster.

**Comment by mgoodness** [comment]
@k8s-bot e2e test this

**Comment by mgoodness** [comment]
@kfox1111 have you had a chance to test the latest iteration? I think it's good to go.

**Comment by mgoodness** [comment]
After further discussion with @prydonius, we've decided not to delete nginx-lego as part of this PR. Instead, I'll create another in which the README and NOTES.txt make it clear that it's been deprecated. There's also [work](https://github.com/kubernetes/helm/issues/1963) planned in the Helm project to support/enforce deprecation.

File: stable/nginx-ingress/Chart.yaml
Comment (prydonius) [inline stable/nginx-ingress/Chart.yaml]: Can we add an icon here? e.g. pointing to https://cdn.wp.nginx.com/wp-content/themes/nginx-theme/assets/img/logo.png?

File: stable/nginx-ingress/README.md
Comment (prydonius) [inline stable/nginx-ingress/README.md]: should this be `controller.stats.service.type`?

File: stable/nginx-ingress/README.md
Comment (prydonius) [inline stable/nginx-ingress/README.md]: defaultBackend.service.type?

**Comment by prydonius** [comment]
@mgoodness just some small fixes in the README, otherwise this lgtm!

**Pull request merged**
